File: modules/ort/src/main/java/org/eclipse/sw360/antenna/ort/workflow/analyzers/OrtResultAnalyzer.java
/*
 * Copyright (c) Bosch Software Innovations GmbH 2019.
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 */
package org.eclipse.sw360.antenna.ort.workflow.analyzers;

import com.here.ort.model.AnalyzerRun;
import com.here.ort.model.OrtResult;
import com.here.ort.model.OutputFormatKt;

import org.eclipse.sw360.antenna.api.exceptions.AntennaException;
import org.eclipse.sw360.antenna.api.workflow.ManualAnalyzer;
import org.eclipse.sw360.antenna.api.workflow.WorkflowStepResult;
import org.eclipse.sw360.antenna.model.artifact.Artifact;
import org.eclipse.sw360.antenna.ort.resolver.OrtResultArtifactResolver;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.IOException;
import java.util.*;
import java.util.stream.Collectors;

public class OrtResultAnalyzer extends ManualAnalyzer {
    private static final Logger LOGGER = LoggerFactory.getLogger(OrtResultAnalyzer.class);

    public OrtResultAnalyzer() {
        this.workflowStepOrder = 700;
    }

    @Override
    public WorkflowStepResult yield() throws AntennaException {
        try {
            return new WorkflowStepResult(createArtifactList(componentInfoFile));
        } catch (IOException e) {
            throw new AntennaException("Error parsing the ORT result file: " + e.getMessage());
        }
    }

    @Override
    public String getName() {
        return "OrtResult";
    }

    Collection<Artifact> createArtifactList(File ortResultFile) throws IOException {
        LOGGER.debug("Creating artifact list from ORT result file '" + ortResultFile + "'.");
        OrtResult result = OutputFormatKt.mapper(ortResultFile).readValue(ortResultFile, OrtResult.class);
        OrtResultArtifactResolver resolver = new OrtResultArtifactResolver(result);
        AnalyzerRun analyzerRun = Optional.ofNullable(result.getAnalyzer())
                .orElseThrow(() -> new IOException("No analyzer run found in ORT result file."));
        return analyzerRun.getResult().getPackages().stream()
                .map(p -> resolver.apply(p.getPkg()))
                .collect(Collectors.toSet());
    }
}

File: modules/ort/src/main/kotlin/workflow/analyzers/OrtResultAnalyzer.kt
/*
 * Copyright (c) Bosch Software Innovations GmbH 2019.
 * Copyright (c) Bosch.IO GmbH 2020.
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 */
package org.eclipse.sw360.antenna.ort.workflow.analyzers

import org.ossreviewtoolkit.model.OrtResult
import org.ossreviewtoolkit.model.readValue

import org.eclipse.sw360.antenna.api.workflow.ManualAnalyzer
import org.eclipse.sw360.antenna.api.workflow.WorkflowStepResult
import org.eclipse.sw360.antenna.model.artifact.Artifact
import org.eclipse.sw360.antenna.ort.resolver.OrtResultArtifactResolver

import org.slf4j.LoggerFactory

import java.io.File
import java.io.IOException

class OrtResultAnalyzer : ManualAnalyzer() {
    companion object {
        private val LOGGER = LoggerFactory.getLogger(OrtResultAnalyzer::class.java)
    }

    private val keepExcludedArtifactsProperty = "keep.excluded.artifacts"
    private var omitExcludedArtifacts = true

    init {
        workflowStepOrder = 700
    }

    override fun yield() = WorkflowStepResult(createArtifactList(componentInfoFile))

    override fun getName() = "OrtResult"

    override fun configure(configMap: Map<String, String>) {
        super.configure(configMap)
        omitExcludedArtifacts = !getBooleanConfigValue(keepExcludedArtifactsProperty, configMap)
    }

    @Throws(IOException::class)
    fun createArtifactList(ortResultFile: File): Collection<Artifact> {
        LOGGER.debug("Creating artifact list from ORT result file '$ortResultFile'.")

        val result = ortResultFile.readValue<OrtResult>()
        if (result.analyzer == null) throw IOException("No analyzer run found in ORT result file.")

        val resolver = OrtResultArtifactResolver(result)
        return result.getPackages(omitExcludedArtifacts).map { (pkg, _) -> resolver.apply(pkg) }.toSet()
    }
}
